WikiWatershed/model-my-watershed
# BiG-CZ Middleware

## Overview

Previously we were setting BiG-CZ specific settings based on an environment variable. However, since in live deployment both MMW and BiG-CZ will be running off the same (virtual) servers, we cannot rely on environment variables to distinguish them.

This approach looks for a `?bigcz=true` query parameter. If found, it sets a session variable which turns on all relevant switches. These settings remain until the user’s session expires, or they pass in `?bigcz=false` which turns them off.

In addition, we also inspect the request’s `HTTP_HOST` header, and if it matches the BiG-CZ pattern we turn these settings on. If it matches the MMW pattern we turn these settings off.

Also tagging @mmcfarland and @hectcastro for review.

Connects #1904 

### Notes

Since this is an additional middleware, it will ever-so-slightly slow down _all_ requests.

The query strings are for development support. They will not work on the canonical URLs, i.e. you cannot use `?bigcz=true` on `staging.app.wikiwatershed.org` and have it be BiG-CZ, or vice-versa. The `HTTP_HOST` will trump the query parameter every time.

For the `HTTP_HOST` cases, we're setting the session variable _in every request_. This is redundant, but since each request is completely independent, I couldn't think of a way to not do this.

## Testing Instructions

 * Check out this branch
 * Go to [:8000/?bigcz=true](http://localhost:8000/?bigcz=true) and type in `clientSettings.data_catalog_enabled` in the console. Ensure its value is `true`.
 * Go to [:8000](http://localhost:8000/) without the query parameter, and ensure that `data_catalog_enabled` is still `true`
 * Got to [:8000?bigcz=false](http://localhost:8000/?bigcz=false) and ensure that `data_catalog_enabled` is now `false`
 * Go to [:8000](http://localhost:8000/) without the query parameter, and ensure that `data_catalog_enabled` is still `false`
* Try to spoof the `Host` header to `portal.bigcz.org` or `staging.portal.bigcz.org` and ensure that `daata_catalog_enabled` is `true`
